src/server/Archive.ts

```diff
@@ -140,18 +140,62 @@ export async function readGameRecord(
       log.error(
         `${gameId}: Error reading game record from R2. Non-Error type: ${String(error)}`,
       );
-      return null;
+    } else {
+      const { message, stack, name } = error;
+      // Log the error for monitoring purposes
+      log.error(`${gameId}: Error reading game record from R2: ${error}`, {
+        message: message,
+        stack: stack,
+        name: name,
+        ...(error && typeof error === "object" ? error : {}),
+      });
     }
-    const { message, stack, name } = error;
-    // Log the error for monitoring purposes
-    log.error(`${gameId}: Error reading game record from R2: ${error}`, {
-      message: message,
-      stack: stack,
-      name: name,
-      ...(error && typeof error === "object" ? error : {}),
+    return readGameRecordFallback(gameId);
+  }
+}
+
+export async function readGameRecordFallback(
+  gameId: GameID,
+): Promise<GameRecord | null> {
+  try {
+    const response = await fetch(config.replayFallbackUrl(gameId), {
+      headers: {
+        Accept: "application/json",
+      },
     });
 
-    // Return null instead of throwing the error
+    if (!response.ok) {
+      throw new Error(
+        `Http error: non-successful http status ${response.status}`,
+      );
+    }
+
+    const contentType = response.headers.get("Content-Type")?.split(";")?.[0];
+    if (contentType !== "application/json") {
+      throw new Error(
+        `Http error: unexpected content type "${response.headers.get("Content-Type")}"`,
+      );
+    }
+
+    return await response.json();
```

Comment:
This implicity `any` is doing an unsafe cast here. We need to validate the schema with zod. Something like this should work:
```suggestion
    const json = await response.json();
    const parsed = GameRecordSchema.parse(json);
    if (!parsed.success) {
      const error = z.prettifyError(parsed.error);
      log.error(`${gameId}: Error parsing game record: ${error}`);
      return null;
    }
    return parsed.result;
```

Reply:
I think this should do the trick:
https://github.com/aaa4xu/OpenFrontIO/pull/1